### PR TITLE
improve schema serialization in BinarySerializer

### PR DESCRIPTION
**Pull request by Linary**
Improve #248

Change-Id: Ieb0734f21c7d1cda3a3f586048109da048c307d7

### CODE REVIEW

File: hugegraph-core/src/main/java/com/baidu/hugegraph/backend/serializer/BinarySerializer.java [hunk 1]
```diff
@@ -812,96 +820,291 @@ protected static boolean indexFieldValuesUnmatched(byte[] value,
...
-            String value = text.column(name);
-            bin.column(joinIdKey(id, name),
-                       StringEncoding.encode(value));
-        }
-        return bin;
+    @Override
+    public BackendEntry writeEdgeLabel(EdgeLabel edgeLabel) {
+        BinaryBackendEntry entry = newBackendEntry(edgeLabel);
+        Id id = edgeLabel.id();
+        entry.column(joinIdKey(id, HugeKeys.NAME),
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/backend/serializer/BinarySerializer.java]: add method writeColumn(Id id, HugeKeys key, String value)

File: hugegraph-core/src/main/java/com/baidu/hugegraph/backend/serializer/BinarySerializer.java [hunk 2]
```diff
@@ -812,96 +820,291 @@ protected static boolean indexFieldValuesUnmatched(byte[] value,
...
+                     StringEncoding.encode(edgeLabel.name()));
+        entry.column(joinIdKey(id, HugeKeys.SOURCE_LABEL),
+                     writeId(edgeLabel.sourceLabel()));
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/backend/serializer/BinarySerializer.java]: add method writeColumn(Id id, HugeKeys key, Id value)

File: hugegraph-core/src/main/java/com/baidu/hugegraph/backend/serializer/BinarySerializer.java [hunk 3]
```diff
@@ -812,96 +820,291 @@ protected static boolean indexFieldValuesUnmatched(byte[] value,
...
-        }
-        return text;
-    }
 
-    // TODO: improve schema serialize
-    private final TextSerializer textSerializer = new TextSerializer();
+        BinaryBackendEntry entry = this.convertEntry(backendEntry);
+        Id id = entry.id().origin();
+        String name = StringEncoding.decode(readColumn(entry, id,
+                                                       HugeKeys.NAME));
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/backend/serializer/BinarySerializer.java]: add method readColumnString() and readColumnId()

File: hugegraph-core/src/main/java/com/baidu/hugegraph/backend/serializer/BinarySerializer.java [hunk 4]
```diff
@@ -812,96 +820,291 @@ protected static boolean indexFieldValuesUnmatched(byte[] value,
...
+        return ids;
+    }
+
+    private static byte[] writeEnum(SerialEnum serialEnum) {
+        return new byte[]{serialEnum.code()};
+    }
+
+    private static <T extends SerialEnum> T readEnum(byte[] value,
+                                                     Class<T> clazz) {
+        return SerialEnum.fromCode(clazz, value[0]);
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/backend/serializer/BinarySerializer.java]: check value length

File: hugegraph-core/src/main/java/com/baidu/hugegraph/backend/serializer/BinarySerializer.java [hunk 5]
```diff
@@ -812,96 +820,291 @@ protected static boolean indexFieldValuesUnmatched(byte[] value,
...
+
+        EdgeLabel edgeLabel = new EdgeLabel(graph, id, name);
+        edgeLabel.sourceLabel(readId(sourceLabel));
+        edgeLabel.targetLabel(readId(targetLabel));
+        edgeLabel.frequency(readEnum(frequency, Frequency.class));
+        edgeLabel.properties(readIds(properties));
+        edgeLabel.sortKeys(readIds(sortKeys));
+        edgeLabel.nullableKeys(readIds(nullablekeys));
+        edgeLabel.indexLabels(readIds(indexLabels));
+        edgeLabel.enableLabelIndex(enableLabelIndex[0] != 0);
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/backend/serializer/BinarySerializer.java]: check enableLabelIndex length and add method readColumnBool

File: hugegraph-core/src/main/java/com/baidu/hugegraph/backend/serializer/BinarySerializer.java [hunk 6]
```diff
@@ -812,96 +820,324 @@ protected static boolean indexFieldValuesUnmatched(byte[] value,
...
+        for (int i = 0; i < size; i++) {
+            Id id = buffer.readId();
+            ids[i] = id;
+        }
+        return ids;
+    }
+
+    private static byte[] readColumn(BinaryBackendEntry entry,
+                                     Id id, HugeKeys key) {
+        return entry.column(joinIdKey(id, key)).value;
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/backend/serializer/BinarySerializer.java]: check column() return null

File: hugegraph-core/src/main/java/com/baidu/hugegraph/backend/serializer/BinarySerializer.java [hunk 7]
```diff
@@ -812,96 +820,324 @@ protected static boolean indexFieldValuesUnmatched(byte[] value,
...
+        for (Map.Entry<String, Object> e : userdata.entrySet()) {
+            schema.userdata(e.getKey(), e.getValue());
+        }
     }
 
-    @Override
-    public BackendEntry writeIndexLabel(IndexLabel indexLabel) {
-        // TODO Auto-generated method stub
-        return text2bin(this.textSerializer.writeIndexLabel(indexLabel));
+    private static void writeColumn(BinaryBackendEntry entry, Id id,
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/backend/serializer/BinarySerializer.java]: add a inner class SchemaSerializer

File: hugegraph-core/src/main/java/com/baidu/hugegraph/backend/serializer/BinarySerializer.java [hunk 8]
```diff
@@ -812,96 +820,337 @@ protected static boolean indexFieldValuesUnmatched(byte[] value,
...
+            for (int i = 0; i < size; i++) {
+                Id id = buffer.readId();
+                ids[i] = id;
+            }
+            return ids;
+        }
+
+        private byte[] column(HugeKeys key) {
+            BackendColumn column = this.entry.column(formatColumnName(key));
+            E.checkState(column != null, "Invalid key '%s' for entry '%s'",
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/backend/serializer/BinarySerializer.java]: Not found key '%s' from entry %s

File: hugegraph-core/src/main/java/com/baidu/hugegraph/backend/serializer/BinarySerializer.java [hunk 9]
```diff
@@ -812,96 +820,337 @@ protected static boolean indexFieldValuesUnmatched(byte[] value,
...
+            this.entry = entry;
+            Id id = entry.id().origin();
+            String name = readColumnString(HugeKeys.NAME);
+
+            PropertyKey propertyKey = new PropertyKey(graph, id, name);
+            propertyKey.dataType(readColumnEnum(HugeKeys.DATA_TYPE,
+                                                DataType.class));
+            propertyKey.cardinality(readColumnEnum(HugeKeys.CARDINALITY,
+                                                   Cardinality.class));
+            propertyKey.properties(readColumnIds(HugeKeys.PROPERTIES));
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/backend/serializer/BinarySerializer.java]: change to readIds(HugeKeys)

File: hugegraph-core/src/main/java/com/baidu/hugegraph/backend/serializer/BinarySerializer.java [hunk 10]
```diff
@@ -812,96 +820,329 @@ protected static boolean indexFieldValuesUnmatched(byte[] value,
...
+        if (backendEntry == null) {
+            return null;
+        }
+        BinaryBackendEntry entry = this.convertEntry(backendEntry);
+
+        SchemaSerializer serializer = new SchemaSerializer();
+        return serializer.readIndexLabel(graph, entry);
+    }
+
+    private final class SchemaSerializer {
```
[the remark this answers is not in the document]
Comment (Linary) [reply hugegraph-core/src/main/java/com/baidu/hugegraph/backend/serializer/BinarySerializer.java]: used member method `newBackendEntry`

**Pull request merged**